geode-core/src/distributedTest/java/org/apache/geode/pdx/PdxTypeGenerationDUnitTest.java

```diff
@@ -0,0 +1,320 @@
+
+    server1.invoke(() -> {
+      InternalCache cache = ClusterStartupRule.getCache();
+      assertThat(cache).isNotNull();
+
+      // Creating a PdxType from the unmodified JSON document so that the same document can be used
+      // later to confirm that the PdxType associated with it is still registered
+      JSONFormatter.fromJSON(jsonString);
+
+      // Creating more PdxTypes to allow us to confirm that they are all recovered when we attempt
+      // to create a PdxType after restarting the server
+      for (int i = 0; i < addedTypes; ++i) {
+        String replacementField = "counter" + i;
+        String modifiedJSON = jsonString.replace(FIELD_NAME_TO_REPLACE, replacementField);
+        JSONFormatter.fromJSON(modifiedJSON);
+      }
+      PeerTypeRegistration registration =
+          (PeerTypeRegistration) (cache.getPdxRegistry().getTypeRegistration());
+
+      assertThat(registration.getLocalSize()).isEqualTo(addedTypes + 1);
+      assertThat(registration.getTypeToIdSize()).isEqualTo(addedTypes + 1);
+    });
+
+    server1.stop(false);
+    Properties props = new Properties();
+    props.setProperty("log-level", "WARN");
+    int locatorPort1 = locator.getPort();
+    server1 = cluster.startServerVM(1,
+        x -> x.withProperties(props).withConnectionToLocator(locatorPort1));
+
+    server1.invoke(() -> {
+      InternalCache cache = ClusterStartupRule.getCache();
+      assertThat(cache).isNotNull();
+
+      PeerTypeRegistration registration =
+          (PeerTypeRegistration) (cache.getPdxRegistry().getTypeRegistration());
+
+      assertThat(registration.getLocalSize()).isEqualTo(addedTypes + 1);
+      assertThat(registration.getTypeToIdSize()).isEqualTo(0);
+
+      // Attempting to create a PdxType that already exists in the /PdxTypes region in order to
+      // trigger rebuilding of the TypeToId map in PeerTypeRegistration
+      JSONFormatter.fromJSON(jsonString);
+
+      assertThat(registration.getLocalSize()).isEqualTo(addedTypes + 1);
+      assertThat(registration.getTypeToIdSize()).isEqualTo(addedTypes + 1);
+    });
+  }
+
+  @Test
+  public void testNoConflictsWhenGeneratingPdxTypesFromJSONOnMultipleServers() {
+    int repeats = 1000;
+    AsyncInvocation invocation1 = null;
+    AsyncInvocation invocation2 = null;
+
+    invocation1 = server1.invokeAsync(() -> {
+      for (int i = 0; i < repeats; ++i) {
+        JSONFormatter.fromJSON("{\"fieldName" + i + "\": \"value\"}");
+      }
+    });
+    invocation2 = server2.invokeAsync(() -> {
+      for (int i = 0; i < repeats; ++i) {
+        JSONFormatter.fromJSON("{\"fieldName" + i + "\": \"value\"}");
+      }
+    });
+
+    try {
+      invocation1.await();
+      invocation2.await();
+    } catch (Exception ex) {
+      throw new TestException("Exception while awaiting async invocation: " + ex);
+    }
+
+    server1.invoke(() -> {
+      InternalCache cache = ClusterStartupRule.getCache();
+      assertThat(cache).isNotNull();
```

Comment:
Just a preference: I don't think these asserts are necessary.  We would get a NPE instead of an assertion error.